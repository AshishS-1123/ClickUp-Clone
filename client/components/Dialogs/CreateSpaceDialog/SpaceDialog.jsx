import React from "react";
import Dialog from "@mui/material/Dialog";
import CloseIcon from "@mui/icons-material/Close";
import styles from "../dialog.module.css";
import themeColors from "../../../utils/contexts/themeContext";

function CreateSpaceDialog({ open, closeDialog, handleCreateSpace }, ref) {
  return (
    <Dialog
      open={open}
      onBackdropClick={closeDialog}
      sx={{
        "& .MuiDialog-paper": {
          width: "550px",
          height: "419px",
          position: "absolute",
          top: "10px",
          background: themeColors.background,
        },
      }}
    >
      <div className={styles.title_bar}>
        <h1 className={styles.title}>Create new space</h1>
        <CloseIcon
          sx={{
            position: "absolute",
            top: "25px",
            right: "25px",
            color: themeColors.textBoldColor,
            width: "32px",
            height: "32px"
          }}
          onClick={closeDialog}
        />
      </div>
      <div className={styles.content}>
        <label htmlFor="spaceDialog_name" id="dialog_nameLabel" style={{ color: themeColors.textBoldColor }}>Space name</label>
        <input type="text" id="spaceDialog_name" placeholder="Enter space name" ref={ref} />
      </div>
      <button
        className={styles.dialog_button}
        onClick={handleCreateSpace}
        style={{
          background: themeColors.accentColor,
          fontWeight: 700,
        }}
      >
        Create space
      </button>
    </Dialog>
  );
}

export default React.forwardRef(CreateSpaceDialog);
