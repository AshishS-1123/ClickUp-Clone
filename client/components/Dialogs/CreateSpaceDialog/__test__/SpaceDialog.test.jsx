import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SpaceDialog from "../SpaceDialog";

let component = null;

describe("<SpaceDialog /> tests", () => {
  // Snapshot Testing
  it("should match snapshot", () => {
    component = render(
      <SpaceDialog
        open
        closeDialog={() => { }}
        handleCreateSpace={() => { }}
        ref={null}
      />,
    );

    expect(component.container).toMatchSnapshot();
  });

  // Test for syntax errors.
  it("should render without crashing", () => {
    component = render(
      <SpaceDialog
        open
        closeDialog={() => { }}
        handleCreateSpace={() => { }}
        ref={null}
      />,
    );
  });

  //Test that visibility of component can be controlled by open prop.
  it("should be displayed when 'open' prop is passed", () => {
    component = render(
      <SpaceDialog
        open
        closeDialog={() => { }}
        handleCreateSpace={() => { }}
        ref={null}
      />,
    );

    expect(screen.getByRole("dialog")).toBeDefined();
  });

  //Test that visibility of component can be controlled by open prop.
  it("should not be displayed when 'open' prop is passed", () => {
    component = render(
      <SpaceDialog
        open={false}
        closeDialog={() => { }}
        handleCreateSpace={() => { }}
        ref={null}
      />,
    );

    expect(screen.queryByRole("dialog")).toBeNull();
  });

  // Standard behaviour for backdrop. Implementation from mui.
  it("should close when backdrop is clicked", () => {
    const closeDialog = jest.fn();
    const handleCreateSpace = jest.fn();

    component = render(
      <SpaceDialog
        open
        closeDialog={closeDialog}
        handleCreateSpace={handleCreateSpace}
        ref={null}
      />,
    );

    // Get the backdrop.
    const backdrop = screen.getByRole("dialog").parentElement;

    // Click on the backdrop.
    userEvent.click(backdrop);

    // Assert that the closeDialog fundtion was called.
    expect(closeDialog).toHaveBeenCalledTimes(1);
  });

  // Test that function for closing dialog is called when close icon is clicked.
  it("should should close when close icon is clicked", () => {
    const closeDialog = jest.fn();
    const handleCreateSpace = jest.fn();

    component = render(
      <SpaceDialog
        open
        closeDialog={closeDialog}
        handleCreateSpace={handleCreateSpace}
        ref={null}
      />,
    );

    // Get the close icon button.
    const button = screen.getByTestId("CloseIcon");

    // Click on the button.
    userEvent.click(button);

    // Assert that the closeDialog fundtion was called.
    expect(closeDialog).toHaveBeenCalledTimes(1);
  });

  // Check UI layout.
  it("should have a title, close icon, input element and submit button", () => {
    component = render(
      <SpaceDialog
        open
        closeDialog={jest.fn()}
        handleCreateSpace={jest.fn()}
        ref={null}
      />,
    );

    // Check label is present.
    expect(screen.getByLabelText("Space name")).toBeDefined();
    // Check input is present.
    expect(screen.getByPlaceholderText("Enter space name")).toBeDefined();
    // Check submit button is present
    expect(screen.getByText("Create space")).toBeDefined();
  });
});
