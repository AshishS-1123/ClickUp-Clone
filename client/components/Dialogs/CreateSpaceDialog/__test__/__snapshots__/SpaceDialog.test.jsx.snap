// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceDialog /> tests should match snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
